src/main/java/com/amazon/opendistroforelasticsearch/sql/query/AggregationQueryAction.java

```diff
@@ -231,10 +238,34 @@ private AggregationBuilder getGroupAgg(Field field, Select select2) throws SqlPa
         }
 
         if (!refrence) lastAgg = aggMaker.makeGroupAgg(field);
-        
+
+        // find if we have order for that aggregation. As of now only special case for script fields
+        if (field.isScriptField()) {
+            addOrderByScriptFieldIfPresent((ScriptMethodField) field, select, lastAgg);
+        }
+
         return lastAgg;
     }
 
+    private void addOrderByScriptFieldIfPresent(ScriptMethodField field, Select select, AggregationBuilder lastAgg) {
+        // This is a hacky way, but it's the best that could be done now.
+
+        // TODO: Explore other ways to correlate different fields/functions in the query (params?)
+        ScriptMethodField groupByScriptField = field;
+        SQLExpr groupByExpression = groupByScriptField.getExpression();
+        for (Order order : select.getOrderBys()) {
+            if (order.isScript()) {
+                ScriptMethodField orderByScriptField = (ScriptMethodField) order.getSortField();
+                SQLExpr orderByExpression = orderByScriptField.getExpression();
+
+                if (groupByExpression.equals(orderByExpression)) {
```

Comment:
I forgot the intent of recent added `expression` in `Field` and have the following questions:

 1. What's information missing in `ScriptMethodField` to force you keep the reference to the expr in original AST?
 2. Does the equality of expression imply the equality of `ScriptMethodField`?

Reply:
1. Script method field had only generated script information, and due to generated field names inside of the script the code can't understand that those generated scripts are actually related to the same expression.
2. No, script method field has more information, for example field alias

Reply:
Thanks for clarification.